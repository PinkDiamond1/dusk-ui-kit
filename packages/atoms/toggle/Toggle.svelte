<script>
  import types from "@dusk-network/helpers/types.js";
  import variants from "@dusk-network/helpers/variants.js";
  import Icon from "@dusk-network/icon/Icon.svelte";
  import "./styles.css";

  /**
   * Sets the variant of the Toggle.
   * @type {"brand" | "cta" | "info" | "success" | "warning" | "danger" }
   */
  export let variant = variants.TOGGLE.BRAND;

  /**
   * Sets the type of the Toggle.
   * @type { "checkbox" | "radio" }
   */
  export let type = types.TOGGLE.SWITCH;

  /**
   * Sets the `name` attribute on the Toggle.
   */
  export let name = "";

  /**
   * Sets the `checked` attrtibute on the Toggle when true.
   */
  export let checked = false;

  /**
   * Sets the `id` of the Toggle if set, otherwise the ID is generated.
   */
  export let id = "__DUK-toggle" + Math.random().toString(36);

  /**
   * Used as a switch when the Toggle is clicked.
   */
  export let value = false;

  /**
   * Used to bind the Toggle to a group.
   */
  export let group = false;

  /**
   * Sets the `disabled` attribute on the Toggle when true.
   */
  export let disabled = false;

  /**
   * Sets the selected state of the Toggle when true.
   */
  export let selected = false;

  /**
   * Used to set the On icon of the Toggle switch thumb.
   */
  export let onIcon = "";

  /**
   * Used to set the Off icon of the Toggle switch thumb.
   */
  export let offIcon = "";
</script>

<div
  class="{$$props.class || ''} duk-toggle"
  class:duk-toggle--brand="{variant === variants.TOGGLE.BRAND}"
  class:duk-toggle--cta="{variant === variants.TOGGLE.CTA}"
  class:duk-toggle--info="{variant === variants.TOGGLE.INFO}"
  class:duk-toggle--success="{variant === variants.TOGGLE.SUCCESS}"
  class:duk-toggle--warning="{variant === variants.TOGGLE.WARNING}"
  class:duk-toggle--danger="{variant === variants.TOGGLE.DANGER}"
>
  {#if type === types.TOGGLE.CHECKBOX}
    <label class="duk-toggle__label duk-toggle__label--checkbox" for="{id}">
      <input
        class="duk-toggle__checkbox"
        type="checkbox"
        role="checkbox"
        name="{name}"
        value="{value}"
        id="{id}"
        disabled="{disabled || undefined}"
        bind:checked
        on:click="{() => (value = !value)}"
        on:click
        on:change
      />
      <span class="duk-toggle__checkbox-label-wrapper">
        <slot />
      </span>
    </label>
  {:else if type === types.TOGGLE.BUTTON}
    <input
      class="duk-toggle__button"
      type="radio"
      role="radio"
      name="{name}"
      value="{value}"
      id="{id}"
      disabled="{disabled || undefined}"
      selected="{selected || undefined}"
      bind:group
      on:click
      on:change
    />
    <label class="duk-toggle__label duk-toggle__label--button" for="{id}">
      <slot />
    </label>
  {:else if type === types.TOGGLE.RADIO}
    <label class="duk-toggle__label duk-toggle__label--radio" for="{id}">
      <input
        class="duk-toggle__radio"
        type="radio"
        role="radio"
        name="{name}"
        value="{value}"
        id="{id}"
        disabled="{disabled || undefined}"
        selected="{selected || undefined}"
        bind:group
        on:click
        on:change
      />
      <span class="duk-toggle__radio-label-wrapper"> <slot /></span>
    </label>
  {:else}
    <div class="duk-toggle__switch-wrapper">
      <input
        class="duk-toggle__switch"
        type="checkbox"
        role="switch"
        name="{name}"
        value="{value}"
        id="{id}"
        disabled="{disabled || undefined}"
        data-action="aria-switch"
        aria-labelledby="{id}-label"
        aria-checked="{checked ? 'true' : 'false'}"
        bind:checked
        on:click="{() => (value = !value)}"
        on:click
        on:change
      />
      <label for="{id}" class="duk-toggle__switch-track"></label>
      <label for="{id}" class="duk-toggle__switch-thumb">
        {#if value && onIcon}
          <Icon name="{onIcon}" />
        {:else if offIcon}
          <Icon name="{offIcon}" />
        {/if}
      </label>
      <label class="duk-toggle__label duk-toggle__label--switch" for="{id}">
        <slot />
      </label>
    </div>
  {/if}
</div>
