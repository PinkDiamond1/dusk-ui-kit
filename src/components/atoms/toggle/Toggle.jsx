/* @flow */
import * as React from 'react'
import ReactToggle from 'react-toggle'

import './Toggle.scss'

export const ToggleVariants = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  INFO: 'info',
  SUCCESS: 'success',
  WARNING: 'warning',
  DANGER: 'danger',
}

type Props = {
  className: string,
  value: string,
  onChange: Function,
  variant: 'primary' | 'secondary' | 'info' | 'success' | 'warning' | 'danger',
  disabled: boolean,
  ariaLabel?: string,
  name?: string,
  id?: string,
  defaultChecked?: boolean,
}

const Toggle = (props: Props): React.Element<*> => {
  const {
    className,
    value,
    onChange,
    variant,
    disabled,
    ariaLabel,
    name,
    id,
    defaultChecked,
  } = props
  let updatedClassName = `__duk-button-toggle ${className}`
  if (variant) updatedClassName += `__duk-button-toggle--${variant}`
  return (
    <ReactToggle
      className={updatedClassName}
      value={value}
      onChange={onChange}
      aria-label={ariaLabel}
      name={name}
      id={id}
      defaultChecked={defaultChecked}
      disabled={disabled}
      icons={false}
    />
  )
}

Toggle.defaultProps = {
  className: '',
  variant: ToggleVariants.PRIMARY,
  onToggle: () => {},
  disabled: false,
  ariaLabel: '',
  name: '',
  id: '',
  value: '',
  defaultChecked: false,
}

export default Toggle
