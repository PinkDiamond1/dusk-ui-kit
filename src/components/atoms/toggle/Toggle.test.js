import React from 'react'
import renderer from 'react-test-renderer'

import Toggle from './Toggle'

it('test to see if the Toggle renders correctly', () => {
  const tree = renderer.create(<Toggle />).toJSON()
  expect(tree).toMatchSnapshot()
})
