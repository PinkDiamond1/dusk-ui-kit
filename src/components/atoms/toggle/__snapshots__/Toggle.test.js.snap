// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to see if the Toggle renders correctly 1`] = `
<div
  className="react-toggle __duk-button-toggle __duk-button-toggle--primary"
  onClick={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="react-toggle-track"
  >
    <div
      className="react-toggle-track-check"
    />
    <div
      className="react-toggle-track-x"
    />
  </div>
  <div
    className="react-toggle-thumb"
  />
  <input
    aria-label=""
    className="react-toggle-screenreader-only"
    defaultChecked={false}
    disabled={false}
    id=""
    name=""
    onBlur={[Function]}
    onFocus={[Function]}
    type="checkbox"
    value=""
  />
</div>
`;
